### ISSUE TITLE
Understand performance regression issues

### ISSUE THREAD

**Issue by elmattic** [elmattic/navi#10]
Going from 86dd5497440773561c1c6dcf5a3be7ae373c248e to 553cd51fc8fa2b83cc5fdbeab9bd52744a84a3c8 we can observe performance regression issues. This could be due to vm.o sheer code size increase (16028 bytes to 16852) or code layout that changed.

bench | from | to
------|------|---
6 x MLOAD | 4.9s | 7.1s

**Comment by elmattic** [comment]
I've been able to get back to previous performance level by two means:

1. Merge of some PUSH opcodes
2. Building the VM with PGO

I also tried to re-organize code layout by hand using opcode frequency (thanks to Alethio):
```
push1:
..
push2:
..
swap1:
..
jumpdest:
..
```

That did not help. I suspect Clang to move things around anyway. My guess is that PGO can lead to a better Hot/Cold code partitioning. Of course understanding this regression would need further work, like using perf, etc.

**But PGO looks really promising.**
